test/test-lint.cpp

```diff
@@ -2811,6 +2811,201 @@ TEST(
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+
+  {
+    // function f() {
+    //   x;
+    //   x = 10;
+    //   {
+    //     eval("var x = 42;");
+    //   }
+    //   x;
+    //   x = 10;
+    // }
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_enter_block_scope();
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_exit_block_scope();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors, IsEmpty());
+  }
+  {
+    // function f() {
+    //   eval("var x = 42;");
+    // }
+    // function g() {
+    //   x;
+    // }
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_exit_function_scope();
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors,
+                ElementsAre(ERROR_TYPE_FIELD(error_use_of_undeclared_variable,
+                                             name, span_matcher(use))));
+  }
+
+  {
+    // function f() {
+    //   eval("var x = 42;");
+    // }
+    // x;
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_exit_function_scope();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors,
+                ElementsAre(ERROR_TYPE_FIELD(error_use_of_undeclared_variable,
+                                             name, span_matcher(use))));
+  }
+
+  {
+    // function f() {
+    //   function g() {
+    //     eval("var x = 42;");
+    //   }
+    //   x;
+    //   x = 10;
+    // }
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_exit_function_scope();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_end_of_module();
+
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(ERROR_TYPE_FIELD(error_use_of_undeclared_variable, name,
+                                     span_matcher(use)),
+                    ERROR_TYPE_FIELD(error_assignment_to_undeclared_variable,
+                                     assignment, span_matcher(use))));
+  }
+}
 }
```

Comment:
More silencing test cases I thought of:
```js
function f(a = eval('var x = 42;')) {
  console.log(x); // no error
}
f();
console.log(x); // ERROR (use of undeclared variable)
```

More erroring test cases I thought of:
```js
function f() {
  const x = 42;
  {
    eval("var x = 0;");
    x = 3;  // ERROR (assignment to const variable)
  }
}
f();
```
```js
let eval = console.log;
eval("var x = 42");
x;  // ERROR (use of undeclared variable)
x = 10;  // ERROR (assignment to undeclared variable)
```
```js
let eval = console.log;
function f() {
  eval("var x = 42");
  x;  // ERROR (use of undeclared variable)
  x = 10;  // ERROR (assignment to undeclared variable)
}
f();
```
```js
function f() {
  eval = console.log;
  eval("var x = 42;");
  x;  // ERROR (use of undeclared variable)
  x = 10;  // ERROR (assignment to undeclared variable)
  {
    var eval;
  }
}
f();
```

Even if these erroring tests don't work as desired (i.e. even if the errors are silenced), I think we should still include them to document the current behavior. In the future, when we want to fix the false negatives, we can tweak the already-written tests. (The test on line 2889 falls in this bucket IMO.) If we do include them, I'd create another `TEST` so their false-negative-ness is more apparent.